docs/JAX_porting_PyTorch_model.md

```diff
@@ -0,0 +1,1607 @@
+y = mod(x)
+y.shape
+```
+
+#### `SwapAxes`, `WindowPartition`, `WindowDepartition` implementations
+
+PyTorch source implementation: https://pytorch.org/vision/stable/_modules/torchvision/models/maxvit.html
+
+```{code-cell} ipython3
+class SwapAxes(nnx.Module):
+    def __init__(self, a: int, b: int):
+        self.a = a
+        self.b = b
+
+    def __call__(self, x: jax.Array) -> jax.Array:
+        res = jnp.swapaxes(x, self.a, self.b)
+        return res
+
+
+class WindowPartition(nnx.Module):
+    def __call__(self, x: jax.Array, p: int) -> jax.Array:
+        # Output array with expected layout of [B, H/P, W/P, P*P, C].
+        B, H, W, C = x.shape
+        P = p
+        # chunk up H and W dimensions
+        x = x.reshape((B, H // P, P, W // P, P, C))
+        x = jnp.permute_dims(x, (0, 1, 3, 2, 4, 5))
+        # colapse P * P dimension
+        x = x.reshape((B, (H // P) * (W // P), P * P, C))
+        return x
+
+
+class WindowDepartition(nnx.Module):
+    def __call__(self, x: jax.Array, p: int, h_partitions: int, w_partitions: int) -> jax.Array:
+        # Output array with expected layout of [B, H, W, C].
+        B, G, PP, C = x.shape
+        P = p
+        HP, WP = h_partitions, w_partitions
+        # split P * P dimension into 2 P tile dimensionsa
+        x = x.reshape((B, HP, WP, P, P, C))
+        # permute into B, HP, P, WP, P, C
+        x = jnp.permute_dims(x, (0, 1, 3, 2, 4, 5))
+        # reshape into B, H, W, C
+        x = x.reshape((B, HP * P, WP * P, C))
+        return x
+```
+
+```{code-cell} ipython3
+x = jnp.ones((3, 4, 5, 6))
+mod = SwapAxes(1, 2)
+y = mod(x)
+assert y.shape == (3, 5, 4, 6)
+
+x = jnp.ones((4, 128, 128, 3))
+mod = WindowPartition()
+y = mod(x, p=16)
+assert y.shape == (4, (128 // 16) * (128 // 16), 16 * 16, 3)
+
+x = jnp.ones((4, (128 // 16) * (128 // 16), 16 * 16, 3))
+mod = WindowDepartition()
+y = mod(x, p=16, h_partitions=128 // 16, w_partitions=128 // 16)
+assert y.shape == (4, 128, 128, 3)
+```
+
+#### `PartitionAttentionLayer` implementation
+
+PyTorch source implementation: https://pytorch.org/vision/stable/_modules/torchvision/models/maxvit.html
+
+```{code-cell} ipython3
+# Layer for partitioning the input tensor into non-overlapping windows and applying attention to each window.
+
+class PartitionAttentionLayer(nnx.Module):
+    def __init__(
+        self,
+        in_channels: int,
+        head_dim: int,
+        # partitioning parameters
+        partition_size: int,
+        partition_type: str,
+        # grid size needs to be known at initialization time
+        # because we need to know hamy relative offsets there are in the grid
+        grid_size: Tuple[int, int],
+        mlp_ratio: int,
+        activation_layer: Callable,
+        norm_layer: Callable[..., nnx.Module],
+        attention_dropout: float,
+        mlp_dropout: float,
+        p_stochastic_dropout: float,
+        rngs: nnx.Rngs = nnx.Rngs(0),
+    ):
```

Comment:
```suggestion
class PartitionAttentionLayer(nnx.Module):
    def __init__(
        self,
        in_channels: int,
        head_dim: int,
        # partitioning parameters
        partition_size: int,
        partition_type: str,
        # grid size needs to be known at initialization time
        # because we need to know hamy relative offsets there are in the grid
        grid_size: Tuple[int, int],
        mlp_ratio: int,
        activation_layer: Callable,
        norm_layer: Callable[..., nnx.Module],
        attention_dropout: float,
        mlp_dropout: float,
        p_stochastic_dropout: float,
        rngs: nnx.Rngs = nnx.Rngs(0),
    ):
    """Layer for partitioning the input tensor into non-overlapping windows and applying attention to each window.
    """
```